File: 백준/Gold/1987. 알파벳/알파벳.cc
#include<bits/stdc++.h>
using namespace std;
const int dy[]={-1,0,1,0};
const int dx[]={0,1,0,-1};
int r,c,ny,nx,cnt;
char a[24][24];
bool visited[26];
void DFS(int y,int x)
{
	cnt=max(cnt,(int)count(visited,visited+26,true));
	for(int i=0;i<4;i++)
	{
		ny = y+dy[i];
		nx = x+dx[i];
		int idx = a[ny][nx]-'A';
		if(ny<0 || ny>=r || nx<0 || nx>=c || visited[idx]) continue;
		if(!visited[idx])
		{
			visited[idx]=true;
			DFS(ny,nx);
			visited[idx]=false;
		}
	}
}
int main()
{
   cin>>r>>c;
   for(int i=0;i<r;i++)
   {
      for(int j=0;j<c;j++)
      {
         cin>>a[i][j];
      }
   }
   visited[a[0][0]-'A']=true;
   DFS(0,0);
   
   cout<<cnt;
}

File: 백준/Gold/1987. 알파벳/알파벳.py
import sys
input = sys.stdin.readline

dy = [-1, 0, 1, 0]
dx = [0, 1, 0, -1]
r, c = map(int, input().split())

a = [list(input().strip()) for _ in range(r)]
visited = [False] * 26
ret = 1

def go(y, x, count):
    global ret
    ret = max(ret, count)

    idx = ord(a[y][x]) - 65
    visited[idx] = True

    for i in range(4):
        ny = y + dy[i]
        nx = x + dx[i]

        if 0 <= ny < r and 0 <= nx < c:
            next_idx = ord(a[ny][nx]) - 65
            if not visited[next_idx]:
                go(ny, nx, count + 1)

    visited[idx] = False


go(0, 0, 1)
print(ret)